### ISSUE TITLE
Plan: /document-question-answering — the one classical model that survived

### ISSUE THREAD

**Issue by bthek1** [bthek1/model_playground#32]
# Plan: `/document-question-answering` — the one classical model that survived

Implements §3.3 of the **Multimodal** roadmap ([`docs/roadmaps/multimodal.md`](https://github.com/bthek1/model_playground/blob/main/docs/roadmaps/multimodal.md)). Independent of #30's engine — this one **does** have a pipeline.

---

## What was checked before planning

- **`Xenova/donut-base-finetuned-docvqa` resolves**, `pipeline_tag: document-question-answering`, `model_type: vision-encoder-decoder` (`VisionEncoderDecoderModel`).
- **4.2.0 really does carry a `document-question-answering` pipeline** — it is in `SUPPORTED_TASKS`. Checked first, because that is the assumption three pages in this repo have now got wrong (roadmap §1). This is the rare case where §8's plain `pipeline()` call is the right answer.
- **The download is ~241 MB at q4f16, not the ~500 MB the roadmap quotes.** The repo totals ~4 GB at fp32 because it publishes **three alternative decoders** (`decoder_model`, `decoder_with_past_model`, `decoder_model_merged`); only the merged one is loaded. Per graph:

  | graph | q4f16 | q8 | fp16 |
  |---|---|---|---|
  | `encoder_model` | 51.5 MB | 91.4 MB | 157.8 MB |
  | `decoder_model_merged` | 189.5 MB | 127.3 MB | 252.9 MB |
  | **total** | **241.0 MB** | **218.7 MB** | **410.7 MB** |

  So `graphs: ["encoder_model", "decoder_model_merged"]` and **measured `bytes`** — a naive sum over the repo would quote a 4 GB model, and a params estimate would miss the three-decoder structure entirely.

- **Neither alternative is buildable.** `impira/layoutlm-document-qa` needs OCR boxes as *input*, which the browser would have to produce first; `stepfun-ai/GOT-OCR-2.0-hf` has no export, and Florence-2 on `/image-to-text` already covers plain OCR.

## Design decisions taken up front

**Resolution is a control, and this page must not inherit §3.1's 512.** A document needs more pixels than a photo before its text is legible to the encoder, so the usual downscale advice is backwards here. `/image-text-to-text` caps at 512 because that is SmolVLM's tile size; Donut's processor has its own input contract and its own trade. Expose the longest edge as a slider, quote what it costs, and let the user trade speed for legibility.

This is written into the plan rather than discovered later because **inheriting a sibling page's hyperparameter is a mistake this repo has already made once**, on `/link-prediction`, where `/graph`'s regularisation cost 0.19 of AUC. Reuse that looks like a decision is often an inheritance.

**Say the privacy argument out loud.** People photograph payslips, medical letters and bank statements. This page answers questions about them without the image leaving the device, which is the reason the whole architecture was chosen — and the one page where a user can feel it. One plain sentence, on the page, not in a tooltip.

**Donut answers extractively and tersely, and that is not a bug to paper over.** It returns a span from the document, not prose. The OUTPUT slot should present it as an answer with its question, and say plainly that the model finds text rather than reasoning about it — the `/video-classification` precedent, where framing the limitation *is* a correctness requirement.

**WebGPU or WASM?** Unlike every model in #30 this one is small enough that the CPU path is real: 219 MB at q8, an encoder plus a short decode. Ship both, let `pickBackend()` decide, and do **not** declare `backends: ["webgpu"]` without measuring first. If WASM turns out to be unusable, the entry says so with a measurement, not a guess.

---

### Phase 1 — the module
- [ ] `src/multimodal/docvqa/types.ts` — catalogue (one entry, `graphs`, measured `bytes`), protocol, the resolution bounds.
- [ ] `engine.ts` — the three engine duties over `pipeline("document-question-answering")`. Warm-up is a real generation, so cap it at a couple of tokens (`vision/caption/engine.ts` is the precedent).
- [ ] `docvqa.worker.ts` + `client.ts`.
- [ ] `hooks/useDocVqa.ts` — the shared contract; the resolution rides in the run payload.

### Phase 2 — the route
- [ ] `routes/document-question-answering.tsx` — four slots, the resolution slider, the privacy sentence, the extractive-answer framing.
- [ ] Document samples (an invoice, a form, a receipt) — reuse `TEXT_SAMPLES` where they fit, add what is missing to `vision/samples.ts`.
- [ ] `REAL_ROUTES`: `"document-question-answering": "/document-question-answering"`.

### Phase 3 — tests
- [ ] Unit: engine, catalogue, hook.
- [ ] Route test, the §8 checklist, plus: **changing the resolution runs nothing** (it is INPUT, like every other pre-run control), and the privacy + extractive copy are asserted, the `/video-classification` precedent.
- [ ] E2E (mocked): zero Hub requests before LOAD.
- [ ] `@slow` spec + `just fe-e2e-docvqa`: a **known answer from a known document** — ask for the total on a sample invoice and assert the number. "Some text appeared" would pass while the processor silently fed the model a blank crop.
- [ ] `model-ids.spec.ts`: add the catalogue to the id check and the dtype/graph check.

### Phase 4 — docs
- [ ] `docs/roadmaps/multimodal.md` §3.3 → **Shipped**, with the corrected size table and the resolution reasoning.
- [ ] `CLAUDE.md` + `.github/copilot-instructions.md` in sync.

---

## Testing

- `just fe-test`, `just fe-lint`, `just fe-build`.
- `just fe-e2e` — the gate.
- `just fe-e2e-models` — the id, and that both graphs publish the requested dtype.
- `just fe-e2e-docvqa` — the known answer, on both backends if WASM proves usable.

## Risks

- **Resolution is the whole correctness surface**, and it fails legibly rather than loudly: too small and the model confidently reads the wrong number. The `@slow` spec asserting a specific figure is the only guard that catches it.
- **`decoder_model_merged` vs the two alternatives.** If the pipeline picks a different decoder than assumed, the quoted bytes are wrong. `just fe-e2e-models` checks the files we *declare*; confirm against a real load once.

**Comment by bthek1** [comment]
## Correction before building: the resolution slider is wrong, and the opposite is right

Checked the processor and the pipeline before writing code. Two of the plan's assumptions hold, one does not.

**Holds — the pipeline can really load it.** `MODEL_FOR_DOCUMENT_QUESTION_ANSWERING_MAPPING_NAMES` maps exactly one entry, `vision-encoder-decoder → VisionEncoderDecoderModel`, which is Donut's model type. `Xenova/donut-base-finetuned-docvqa` is also the pipeline's *own default model*. So §8's plain `pipeline()` call is right, and this stays the one route in the category that does not need its own engine.

**Does not hold — "expose the longest edge as a slider and let the user trade speed for legibility".** `preprocessor_config.json` is `do_resize: true`, `do_thumbnail: true`, `do_pad: true`, `size: { height: 2560, width: 1920 }`, and `thumbnail()` in 4.2.0 reads:

```js
let height = Math.min(input_height, output_height);
let width  = Math.min(input_width,  output_width);
if (height === input_height && width === input_width) return image;  // never upscales
```

It shrinks to fit 2560x1920 preserving aspect, then `do_pad` pads to exactly 2560x1920. **The encoder therefore always sees a 2560x1920 tensor**, whatever it was handed. Consequences:

- **Inference cost is constant.** A slider would save nothing at the model.
- **A smaller source is padded, not upscaled.** Lowering resolution removes real pixels and keeps the compute — legibility traded away for nothing.

So the slider is not a speed/legibility trade; it is a one-way loss. The correct design is the repo's standing rule stated plainly: **never resize for the model** — `AutoProcessor` reads the model's own config and that file *is* the input contract. This page is where that rule bites hardest, and the page should say so, since every other vision route in the repo caps its source and this one deliberately must not.

The only defensible cap is at the processor's own 2560 long edge, purely to bound decode and transfer cost — lossless, because `thumbnail()` would shrink it to exactly there anyway. That is a memory bound, not preprocessing, and it will be commented as such.

The plan's instinct ("a document needs more pixels than a photo") was right; the mechanism was wrong. Recording it here rather than silently building something else — and it is the same lesson the plan itself cites about `/link-prediction`: reuse that looks like a decision is often an inheritance, so run it.

## Two other findings that shape the build

**`answer` can be `null`, and nothing errors.** The pipeline extracts with `decoded.match(/<s_answer>(.*?)<\/s_answer>/)` and returns `[{ answer: null }]` when that misses. So "the model produced nothing usable" is a silent, ordinary outcome that OUTPUT must render explicitly rather than showing a blank panel.

**The prompt format is Donut-specific and hardcoded** in the pipeline (`<s_docvqa><s_question>...</s_question><s_answer>`), which is a further reason this route is one checkpoint rather than a catalogue: a second model would not be driven correctly by this pipeline.

## Dtype

Shipping the default `loadOpts()` — fp16 on WebGPU (410.7 MB), q8 on WASM (218.7 MB), both measured, no pin. `q4f16` would cut WebGPU to 241.0 MB, but pinning it would be a **precaution, not a measurement**, and this repo requires that distinction to be stated. Left for the `@slow` spec to settle, the way `/super-resolution`'s pin was.

**Comment by bthek1** [comment]
**The route this issue shipped has been removed.** Not a defect in the work — it loaded, ran, and its `@slow` spec pinned `us-001` off `invoice.png` — but a size decision taken across the app.

Donut is **411 MB on WebGPU and 596.7 MB on WASM**, and it could never have had a lighter second entry: `DocumentQuestionAnsweringPipeline` hardcodes Donut's own prompt (`<s_docvqa><s_question>…</s_question><s_answer>`), so another architecture would be prompted with tokens it has never seen. A page whose floor is 411 MB, with no second option, is `adding-a-task-page.md` §0 question 2 answered "no" — one step later than it should have been.

Cut alongside `/text-to-audio` (599/1127 MB) and `/image-to-text` (482/544 MB) in 36ba918, with the docs rewritten in e5dcc2b.

**Every finding here is kept, because none of it was about this page:**

- The ORT quantized-decoder bug is the big one. This route is what proved `asrLoadOpts`'s note wrong about itself — *any* encoder-decoder with a quantized decoder fails to open a WASM session on 4.2.0, not just ASR. That now lives in `model/backend.ts` beside the function it generalises.
- `thumbnail()` never upscales, so the resolution slider the plan proposed would have traded legibility for no speed at all.
- `answer` can be `null` and nothing errors.
- It was the one page that *passed* the `SUPPORTED_TASKS` check, which is worth recording next to the three that did not.

Written up in full in `docs/roadmaps/multimodal.md` §3.3, and §0 of `adding-a-task-page.md` now carries the general lesson: **a gate is not a substitute for a size that fits.**
